metis-core/metis-core-service/src/main/java/eu/europeana/metis/core/service/ProxiesService.java

```diff
@@ -394,9 +400,64 @@ public RecordsResponse getListOfFileContentsFromPluginExecution(MetisUserView me
     // Done.
     return new RecordsResponse(records);
   }
+  /**
+   * Get a record from the external resource based on o searchId, workflow execution and {@link PluginType}.
+   *
+   * @param metisUserView the user wishing to perform this operation
+   * @param workflowExecutionId the execution identifier of the workflow
+   * @param pluginType the {@link ExecutablePluginType} that is to be located inside the workflow
+   * @param searchId the searchId of the record we wish to obtain
+   * @return the record from the external resource
+   * @throws GenericMetisException can be one of:
+   * <ul>
+   * <li>{@link eu.europeana.metis.exception.ExternalTaskException} if an error occurred while
+   * retrieving the records from the external resource</li>
+   * <li>{@link eu.europeana.metis.exception.UserUnauthorizedException} if the user is not
+   * authorized to perform this task</li>
+   * <li>{@link eu.europeana.metis.core.exceptions.NoWorkflowExecutionFoundException} if no workflow
+   * execution exists for the provided identifier</li>
+   * </ul>
+   */
+  public Record searchRecordByIdFromPluginExecution(MetisUserView metisUserView, String workflowExecutionId , ExecutablePluginType pluginType, String searchId )
+      throws GenericMetisException {
+
+    // Get the right workflow execution and plugin type.
+    final Pair<WorkflowExecution, ExecutablePlugin> executionAndPlugin = getExecutionAndPlugin(
+        metisUserView, workflowExecutionId, pluginType);
+    if (executionAndPlugin == null) {
+      throw new NoWorkflowExecutionFoundException(String
+          .format("No executable plugin of type %s found for workflowExecution with id: %s",
+              pluginType.name(), workflowExecutionId));
+    }
+
+    // Check whether the searched ID is known as a Europeana ID. Otherwise, assume it is an ecloudId.
+    final String datasetId = executionAndPlugin.getLeft().getDatasetId();
+    String ecloudId = null;
+    try {
+      final String normalizedRecordId = RecordIdUtils.checkAndNormalizeRecordId(datasetId, searchId)
+          .map(id -> RecordIdUtils.composeFullRecordId(datasetId, id)).orElse(null);
+      if (normalizedRecordId != null) {
+        ecloudId = uisClient.getCloudId(ecloudProvider, normalizedRecordId).getId();
+      }
+    } catch (BadContentException e) {
+      // Normalization failed.
+      ecloudId = searchId;
+    } catch (CloudException e) {
+      if (e.getCause() instanceof RecordDoesNotExistException) {
+        // The record ID does not exist.
+        ecloudId = searchId;
+      } else {
+        // Some other connectivity issue.
+        throw new ExternalTaskException(
+            String.format("Failed to lookup cloudId for searchId: %s", searchId), e);
+      }
+    }
+
+    // Try to retrieve the record.
+    return Optional.ofNullable(getRecord(executionAndPlugin.getRight(), ecloudId)).orElse(null);
```

Comment:
Why do `.orElse(null)`? Doesn't `Optional.ofNullable` already handles in case `getRecord` returns null?
This makes it redundant. Either simply return `getRecord` as it is or just use the `Optional.ofNullable`

Reply:
Ah, I think what Adolfo meant here was to check for the nullity of `recordId` itself. I fixed it.